ws/DiscordWebsocket.ts

```diff
@@ -42,13 +42,14 @@ export default class DiscordWebsocket {
     return this;
   }
 
-  disconnect(code?: number) {
-    if (code === 1000 || code === 1001) {
+  disconnect(code?: GATEWAY_CLOSE_EVENT_CODES, reason?: string) {
+    if (code === GATEWAY_CLOSE_EVENT_CODES.NORMAL || code === GATEWAY_CLOSE_EVENT_CODES.GOING_AWAY) {
       this.sessionID = null;
       this._seq = null;
     }
     this._selfDisconnect = true;
-    this.ws?.close(code);
+    this.ws?.off('close', this._onClose);
+    this.ws?.close(code, reason);
```

Comment:
If we reverse them _onClose() would execute and it would think that there was an error
We *want* to disconnect, so we don't need to handle errors, do we?

Reply:
That's a good point
@Khaaz just double check this please in case either of us are going crazy

Reply:
yes that's true, we don't want to loop there. However we might want to call reset on disconnect? Maybe not reset but something that clear all listeners for instance. If we clear only onClose we will have dup onMessage for instance.

Reply:
I tried that and onMessage() still executes one time, it doesn't duplicate

However I noticed that after restarting `{ t: null, s: null, op: 11, d: null }` is being received twice, which means that we are sending the heartbeat twice.  [4128d64](https://github.com/libraryofcode/dis.ts/pull/22/commits/45d919e2eb34b26345fb04e5f3969cc140d9e2bc) fixes that

Reply:
That seems a bit hacky - In this case I'd rather still reverse them here and add `_onClose` cases for 1000/1001 (also 4200 whilst you're at it)

Reply:
that's better